### PR TITLE
[NPU] Run arch35 block-FP8 dense linears on the native MXFP8 GEMM

### PR DESCRIPTION
**Pull request by iridiumine**
## Motivation

On NPU arch35 (A5), plain block-FP8 checkpoints (fp32 128x128 scales, `scale_fmt` unset) fell back to `triton_w8a8_block_fp8_linear` — a CUDA Triton tile GEMM that cannot use the NPU Cube units and runs far slower than the native quantized matmul. The fast native A5 chain (`npu_w8a8_mxfp8_linear`) was only reachable for `scale_fmt="ue8m0"` checkpoints.

## Modifications

- Add `requant_npu_arch35_block_fp8_to_mxfp8` in `w8a8_mxfp8.py`: requantize plain block-FP8 weights once at load time (fp8 payload x block scale -> BF16 -> MXFP8 via `npu_dynamic_mx_quant`), chunked over rows to cap peak memory. The e4m3fn payload is decoded by explicit bit layout to avoid the `aclnnInplaceCopy` error 561103 seen on some CANN builds when converting NPU fp8 tensors directly.
- In `Fp8LinearMethod`, extend the arch35 weight-processing branch to non-ue8m0 checkpoints and remove the triton fallback override in `__init__`, so both checkpoint flavors run the same native A5 GEMM chain.

## Accuracy Tests

```
echo performance | tee /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
sysctl -w vm.swappiness=0
sysctl -w kernel.numa_balancing=0
sysctl -w kernel.sched_migration_cost_ns=50000
export SGLANG_SET_CPU_AFFINITY=1

source /usr/local/Ascend/ascend-toolkit/set_env.sh
source /usr/local/Ascend/nnal/atb/set_env.sh

export HCCL_BUFFSIZE=300
export HCCL_OP_EXPANSION_MODE=AIV
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
export STREAMS_PER_DEVICE=32
export SGLANG_DISAGGREGATION_BOOTSTRAP_TIMEOUT=600
export SGLANG_ENABLE_SPEC_V2=1
export SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1
export ASCEND_USE_FIA=1
export SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=32
export DEEPEP_HCCL_BUFFSIZE=2500
export DEEPEP_NORMAL_LONG_SEQ_ROUND=20
export DEEPEP_NORMAL_LONG_SEQ_PER_ROUND_TOKENS=4096
export HCCL_SOCKET_IFNAME=lo
export GLOO_SOCKET_IFNAME=lo
export HCCL_HOST_SOCKET_PORT_RANGE=auto
export SGLANG_ALLOW_OVERWRITE_LONGER_CONTEXT_LEN=1

sglang serve \
        --model-path $MODEL_PATH \
        --served-model-name $MODEL_PATH \
        --trust-remote-code \
        --attention-backend ascend \
        --device npu \
        --mem-fraction-static 0.92 \
        --tp-size 8 --nnodes 1 --node-rank 0 \
        --chunked-prefill-size 8192 --max-total-tokens 600000 \
        --max-running-requests 32 \
        --moe-a2a-backend deepep --deepep-mode auto \
        --cuda-graph-bs-decode 1 2 4 8 16 \
        --speculative-algorithm DFLASH \
        --speculative-draft-model-path $DFLASH_PATH \
        --speculative-num-draft-tokens 8 \
        --dp-size 2 --enable-dp-attention --enable-dp-lm-head \
```

```
DeprecationWarning: sglang.test.few_shot_gsm8k is deprecated. Use sglang.test.run_eval with eval_name='gsm8k' instead.
  metrics = run_eval(args)
100%|████████████████████████████████████████████████████████| 200/200 [00:49<00:00,  4.01it/s]
Accuracy: 0.970
Invalid: 0.015
Latency: 50.322 s
Output throughput: 463.139 token/s
metrics={'accuracy': np.float64(0.97), 'invalid': np.float64(0.015), 'latency': 50.32184849004261, 'output_throughput': 463.138789597756}
metrics['accuracy']=np.float64(0.97)
```

### Before modification

On NPU A5, plain block-FP8 checkpoints fell back to the `_w8a8_block_fp8_matmul` emulation path (Triton tile GEMM, no Cube unit utilization). A single large-M dense linear call took ~100 ms, which dominated prefill latency (decode is memory-bound on weight reads, so it also benefited but far less).

### After modification

The `_w8a8_block_fp8_matmul` fallback is removed entirely: requantized weights run on the native `npu_w8a8_mxfp8_linear` (Cube unit), eliminating the ~100 ms per-call cost of the emulated GEMM.

End-to-end prefill benchmark on the same workload (128 requests x 16K input tokens, request rate 0.4, max concurrency 32, 1 output token per request):

| Metric                       | Before      | After      |
|------------------------------|-------------|------------|
| Input token throughput       | 924 tok/s   | 6811 tok/s |
| Mean TTFT                    | 477327 ms   | 2901 ms    |
| Median TTFT                  | 535917 ms   | 2556 ms    |
| P99 TTFT                     | 576947 ms   | 7414 ms    |
| Total benchmark duration     | 2216 s      | 301 s      |

i.e. ~7.4x prefill throughput and a ~165x reduction in mean TTFT.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #35303722920](https://github.com/sgl-project/sglang/actions/runs/35303722920)
Latest PR Test (Extra): :x: [Run #35303722776](https://github.com/sgl-project/sglang/actions/runs/35303722776)
Latest PR Test (AMD ROCm 10): :x: [Run #35303722807](https://github.com/sgl-project/sglang/actions/runs/35303722807)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py [hunk 1]
```diff
@@ -18,6 +18,78 @@ def process_npu_arch35_mxfp8_linear_weights(
...
+        raise ValueError(
+            "NPU arch35 MXFP8 linear requires K to be divisible by "
+            f"{2 * group_size}, got {k_dim}."
+        )
+    device = f"npu:{torch.npu.current_device()}"
+    if not weight.is_npu:
+        weight = weight.to(device)
+    scale = layer.weight_scale_inv.data.to(device)
+
+    bf16 = torch.empty(n_dim, k_dim, dtype=torch.bfloat16, device=weight.device)
```
Comment (AndyLi429) [inline python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py]: is it necessary?will cause spus of memory?

Comment (iridiumine) [reply python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py]: Switched to per-row-chunk quantization

File: python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py [hunk 2]
```diff
@@ -18,6 +18,90 @@ def process_npu_arch35_mxfp8_linear_weights(
...
+        e == 0.0,
+        m * (1.0 / 512.0),  # subnormal: m * 2^-9
+        (1.0 + m / 8.0) * torch.pow(2.0, e - 7.0),
+    )
+    # e4m3fn reserves S.1111.111 for NaN (no infinities); max finite is 448.
+    mag = torch.where((e == 15.0) & (m == 7.0), float("nan"), mag)
+    return sign * mag
+
+
+def requant_npu_arch35_block_fp8_to_mxfp8(
```
Comment (iridiumine) [inline python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py]: dsv4  moe method

Comment (iridiumine) [reply python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py]: need e8m0

File: python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py [hunk 3]
```diff
@@ -7,15 +7,103 @@
...
+
+    UE8M0 checkpoints (scales already powers of two) only need re-layout:
+    the 128-group scale can be duplicated to its 1x32 sub-groups exactly.
+    Plain block-FP8 checkpoints carry arbitrary fp32 scales, so the payload
+    must be dequantized and requantized via npu_dynamic_mx_quant instead.
+    """
+    if scale_fmt == "ue8m0":
+        _layout_npu_arch35_ue8m0_weights(layer, weight_block_size)
+    else:
+        requant_npu_arch35_block_fp8_to_mxfp8(layer, weight_block_size)
```
Comment (Hexq0210) [inline python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py]: Unified Naming Conventions

**Review by Hexq0210 — APPROVED** [review]

File: python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py (hunk 3 above)
Comment (iridiumine) [reply python/sglang/srt/hardware_backend/npu/quantization/w8a8_mxfp8.py]: done

**Review by iforgetmyname — COMMENTED** [review]
basically, fp8 per-block quant w/ fp32 scaling factor is different from mxfp8, you can't merge them into one case.
only deepseek's fp8 per-block quant w/ ue8m0 scaling factor is able to convert to mxfp8 as a 128*128 block can be split into sub-blocks of 32 elements sharing the same ue8m0 scaling factor
